File: typescript/packages/fathomdb/src/write-builder.ts
import { BuilderValidationError } from "./errors.js";
import {
  PreserializedJson,
  type ActionInsertInput,
  type ChunkInsertInput,
  type EdgeInsertInput,
  type NodeInsertInput,
  type OperationalAppendInput,
  type OperationalDeleteInput,
  type OperationalPutInput,
  type RunInsertInput,
  type StepInsertInput,
} from "./types.js";

type HandleBase = {
  readonly _builderId: number;
};

let nextBuilderId = 1;

/** Opaque reference to a node added via {@link WriteRequestBuilder}. */
export type NodeHandle = HandleBase & { rowId: string; logicalId: string };

/** Opaque reference to an edge added via {@link WriteRequestBuilder}. */
export type EdgeHandle = HandleBase & { logicalId: string };

/** Opaque reference to a run added via {@link WriteRequestBuilder}. */
export type RunHandle = HandleBase & { id: string };

/** Opaque reference to a step added via {@link WriteRequestBuilder}. */
export type StepHandle = HandleBase & { id: string };

/** Opaque reference to an action added via {@link WriteRequestBuilder}. */
export type ActionHandle = HandleBase & { id: string };

/** Opaque reference to a chunk added via {@link WriteRequestBuilder}. */
export type ChunkHandle = HandleBase & { id: string };

function toJsonString(value: unknown): string {
  if (value instanceof PreserializedJson) return value.json;
  return JSON.stringify(value ?? null);
}

/**
 * Mutable builder that assembles a write request from individual mutations.
 *
 * Handles are returned when adding nodes, edges, runs, steps, actions, and
 * chunks so they can be cross-referenced within the same request. Call
 * {@link build} to produce the finalized wire-format object.
 */
export class WriteRequestBuilder {
  readonly #builderId = nextBuilderId++;
  readonly #request: Record<string, unknown>;

  constructor(label: string) {
    this.#request = {
      label,
      nodes: [],
      node_retires: [],
      edges: [],
      edge_retires: [],
      chunks: [],
      runs: [],
      steps: [],
      actions: [],
      optional_backfills: [],
      vec_inserts: [],
      operational_writes: []
    };
  }

  /**
   * Add a node to the write request.
   *
   * @param input - Node properties including rowId, logicalId, kind, and properties.
   * @returns A handle that can be used to reference this node in edges and chunks.
   */
  addNode(input: NodeInsertInput): NodeHandle {
    const properties = toJsonString(input.properties);
    (this.#request.nodes as Array<Record<string, unknown>>).push({
      row_id: input.rowId,
      logical_id: input.logicalId,
      kind: input.kind,
      properties,
      source_ref: input.sourceRef ?? null,
      upsert: input.upsert ?? false,
      chunk_policy: input.chunkPolicy ?? "preserve",
    });
    return { _builderId: this.#builderId, rowId: input.rowId, logicalId: input.logicalId };
  }

  /**
   * Mark a node as retired (soft-delete) by logical ID or handle.
   *
   * @param logicalId - The node handle or logical ID string to retire.
   * @param sourceRef - Optional provenance source reference.
   */
  retireNode(logicalId: NodeHandle | string, sourceRef?: string): void {
    (this.#request.node_retires as Array<Record<string, unknown>>).push({
      logical_id: this.#resolveNode(logicalId),
      source_ref: sourceRef ?? null
    });
  }

  /**
   * Add an edge connecting two nodes to the write request.
   *
   * @param input - Edge properties including source, target, kind, and properties.
   * @returns A handle that can be used to reference this edge.
   */
  addEdge(input: EdgeInsertInput & { source: NodeHandle | string; target: NodeHandle | string }): EdgeHandle {
    const properties = toJsonString(input.properties);
    (this.#request.edges as Array<Record<string, unknown>>).push({
      row_id: input.rowId,
      logical_id: input.logicalId,
      source_logical_id: this.#resolveNode(input.source),
      target_logical_id: this.#resolveNode(input.target),
      kind: input.kind,
      properties,
      source_ref: input.sourceRef ?? null,
      upsert: input.upsert ?? false,
    });
    return { _builderId: this.#builderId, logicalId: input.logicalId };
  }

  /**
   * Mark an edge as retired (soft-delete) by logical ID or handle.
   *
   * @param logicalId - The edge handle or logical ID string to retire.
   * @param sourceRef - Optional provenance source reference.
   */
  retireEdge(logicalId: EdgeHandle | string, sourceRef?: string): void {
    (this.#request.edge_retires as Array<Record<string, unknown>>).push({
      logical_id: this.#resolveEdge(logicalId),
      source_ref: sourceRef ?? null
    });
  }

  /**
   * Add a text chunk associated with a node.
   *
   * @param input - Chunk properties including id, owning node, and text content.
   * @returns A handle for referencing this chunk in vector inserts.
   */
  addChunk(input: ChunkInsertInput & { node: NodeHandle | string }): ChunkHandle {
    (this.#request.chunks as Array<Record<string, unknown>>).push({
      id: input.id,
      node_logical_id: this.#resolveNode(input.node),
      text_content: input.textContent,
      byte_start: input.byteStart ?? null,
      byte_end: input.byteEnd ?? null,
    });
    return { _builderId: this.#builderId, id: input.id };
  }

  /**
   * Add a run to the write request.
   *
   * @param input - Run properties including id, kind, status, and properties.
   * @returns A handle for referencing this run when adding steps.
   */
  addRun(input: RunInsertInput): RunHandle {
    const properties = toJsonString(input.properties);
    (this.#request.runs as Array<Record<string, unknown>>).push({
      id: input.id,
      kind: input.kind,
      status: input.status,
      properties,
      source_ref: input.sourceRef ?? null,
      upsert: input.upsert ?? false,
      supersedes_id: input.supersedesId ?? null,
    });
    return { _builderId: this.#builderId, id: input.id };
  }

  /**
   * Add a step belonging to a run.
   *
   * @param input - Step properties including id, owning run, kind, status, and properties.
   * @returns A handle for referencing this step when adding actions.
   */
  addStep(input: StepInsertInput & { run: RunHandle | string }): StepHandle {
    const properties = toJsonString(input.properties);
    (this.#request.steps as Array<Record<string, unknown>>).push({
      id: input.id,
      run_id: this.#resolveRun(input.run),
      kind: input.kind,
      status: input.status,
      properties,
      source_ref: input.sourceRef ?? null,
      upsert: input.upsert ?? false,
      supersedes_id: input.supersedesId ?? null,
    });
    return { _builderId: this.#builderId, id: input.id };
  }

  /**
   * Add an action belonging to a step.
   *
   * @param input - Action properties including id, owning step, kind, status, and properties.
   * @returns A handle for referencing this action.
   */
  addAction(input: ActionInsertInput & { step: StepHandle | string }): ActionHandle {
    const properties = toJsonString(input.properties);
    (this.#request.actions as Array<Record<string, unknown>>).push({
      id: input.id,
      step_id: this.#resolveStep(input.step),
      kind: input.kind,
      status: input.status,
      properties,
      source_ref: input.sourceRef ?? null,
      upsert: input.upsert ?? false,
      supersedes_id: input.supersedesId ?? null,
    });
    return { _builderId: this.#builderId, id: input.id };
  }

  /**
   * Queue an optional projection backfill task (e.g. FTS or vector).
   *
   * @param target - Which projection to backfill (`"fts"`, `"vec"`, or `"all"`).
   * @param payload - Projection-specific payload data.
   */
  addOptionalBackfill(target: "fts" | "vec" | "all", payload: unknown): void {
    (this.#request.optional_backfills as Array<Record<string, unknown>>).push({
      target,
      payload: toJsonString(payload),
    });
  }

  /**
   * Add a vector embedding associated with a chunk.
   *
   * @param input - The chunk reference and embedding vector.
   */
  addVecInsert(input: { chunk: ChunkHandle | string; embedding: number[] }): void {
    (this.#request.vec_inserts as Array<Record<string, unknown>>).push({
      chunk_id: this.#resolveChunk(input.chunk),
      embedding: input.embedding
    });
  }

  /**
   * Append a mutation to an operational collection.
   *
   * @param input - Append input including collection name, record key, and payload.
   */
  addOperationalAppend(input: OperationalAppendInput): void {
    const payloadJson = toJsonString(input.payloadJson);
    (this.#request.operational_writes as Array<Record<string, unknown>>).push({
      type: "append",
      collection: input.collection,
      record_key: input.recordKey,
      payload_json: payloadJson,
      source_ref: input.sourceRef ?? null,
    });
  }

  /**
   * Put (upsert) a record into an operational collection.
   *
   * @param input - Put input including collection name, record key, and payload.
   */
  addOperationalPut(input: OperationalPutInput): void {
    const payloadJson = toJsonString(input.payloadJson);
    (this.#request.operational_writes as Array<Record<string, unknown>>).push({
      type: "put",
      collection: input.collection,
      record_key: input.recordKey,
      payload_json: payloadJson,
      source_ref: input.sourceRef ?? null,
    });
  }

  /**
   * Delete a record from an operational collection.
   *
   * @param input - Delete input including collection name and record key.
   */
  addOperationalDelete(input: OperationalDeleteInput): void {
    (this.#request.operational_writes as Array<Record<string, unknown>>).push({
      type: "delete",
      collection: input.collection,
      record_key: input.recordKey,
      source_ref: input.sourceRef ?? null,
    });
  }

  /**
   * Resolve all handles and produce the finalized write request object.
   *
   * @returns A deep clone of the assembled wire-format request.
   * @throws {BuilderValidationError} If any handle belongs to a different builder.
   */
  build(): Record<string, unknown> {
    return structuredClone(this.#request);
  }

  #assertOwnership(handle: HandleBase): void {
    if (handle._builderId !== this.#builderId) {
      throw new BuilderValidationError("handle belongs to a different WriteRequestBuilder");
    }
  }

  #resolveNode(handleOrId: NodeHandle | string): string {
    if (typeof handleOrId === "string") {
      return handleOrId;
    }
    this.#assertOwnership(handleOrId);
    return handleOrId.logicalId;
  }

  #resolveEdge(handleOrId: EdgeHandle | string): string {
    if (typeof handleOrId === "string") {
      return handleOrId;
    }
    this.#assertOwnership(handleOrId);
    return handleOrId.logicalId;
  }

  #resolveRun(handleOrId: RunHandle | string): string {
    if (typeof handleOrId === "string") {
      return handleOrId;
    }
    this.#assertOwnership(handleOrId);
    return handleOrId.id;
  }

  #resolveStep(handleOrId: StepHandle | string): string {
    if (typeof handleOrId === "string") {
      return handleOrId;
    }
    this.#assertOwnership(handleOrId);
    return handleOrId.id;
  }

  #resolveChunk(handleOrId: ChunkHandle | string): string {
    if (typeof handleOrId === "string") {
      return handleOrId;
    }
    this.#assertOwnership(handleOrId);
    return handleOrId.id;
  }
}

File: archive/pre-0.6.0/python/fathomdb/_write_builder.py
from __future__ import annotations

from dataclasses import dataclass
from itertools import count
from typing import Any

from ._types import (
    ActionInsert,
    ChunkInsert,
    ChunkPolicy,
    EdgeInsert,
    EdgeRetire,
    NodeInsert,
    NodeRetire,
    OperationalAppend,
    OperationalDelete,
    OperationalPut,
    OptionalProjectionTask,
    ProjectionTarget,
    RunInsert,
    StepInsert,
    VecInsert,
    WriteRequest,
)
from .errors import BuilderValidationError

_builder_ids = count(1)


@dataclass(frozen=True)
class NodeHandle:
    """Opaque reference to a node added via WriteRequestBuilder."""

    row_id: str
    logical_id: str
    _builder_id: int


@dataclass(frozen=True)
class EdgeHandle:
    """Opaque reference to an edge added via WriteRequestBuilder."""

    logical_id: str
    _builder_id: int


@dataclass(frozen=True)
class RunHandle:
    """Opaque reference to a run added via WriteRequestBuilder."""

    id: str
    _builder_id: int


@dataclass(frozen=True)
class StepHandle:
    """Opaque reference to a step added via WriteRequestBuilder."""

    id: str
    _builder_id: int


@dataclass(frozen=True)
class ActionHandle:
    """Opaque reference to an action added via WriteRequestBuilder."""

    id: str
    _builder_id: int


@dataclass(frozen=True)
class ChunkHandle:
    """Opaque reference to a chunk added via WriteRequestBuilder."""

    id: str
    _builder_id: int


@dataclass(slots=True)
class _PendingEdge:
    row_id: str
    logical_id: str
    source: NodeHandle | str
    target: NodeHandle | str
    kind: str
    properties: Any
    source_ref: str | None
    upsert: bool


@dataclass(slots=True)
class _PendingNodeRetire:
    logical_id: NodeHandle | str
    source_ref: str | None


@dataclass(slots=True)
class _PendingEdgeRetire:
    logical_id: EdgeHandle | str
    source_ref: str | None


@dataclass(slots=True)
class _PendingChunk:
    id: str
    node: NodeHandle | str
    text_content: str
    byte_start: int | None
    byte_end: int | None
    content_hash: str | None


@dataclass(slots=True)
class _PendingStep:
    id: str
    run: RunHandle | str
    kind: str
    status: str
    properties: Any
    source_ref: str | None
    upsert: bool
    supersedes_id: str | None


@dataclass(slots=True)
class _PendingAction:
    id: str
    step: StepHandle | str
    kind: str
    status: str
    properties: Any
    source_ref: str | None
    upsert: bool
    supersedes_id: str | None


@dataclass(slots=True)
class _PendingVec:
    chunk: ChunkHandle | str
    embedding: list[float]


class WriteRequestBuilder:
    """Mutable builder that assembles a WriteRequest from individual mutations.

    Handles are returned when adding nodes, edges, runs, steps, actions, and
    chunks so they can be cross-referenced within the same request.  Call
    :meth:`build` to produce the final :class:`WriteRequest`.
    """

    def __init__(self, label: str) -> None:
        self._builder_id = next(_builder_ids)
        self._label = label
        self._nodes: list[NodeInsert] = []
        self._node_retires: list[_PendingNodeRetire] = []
        self._edges: list[_PendingEdge] = []
        self._edge_retires: list[_PendingEdgeRetire] = []
        self._chunks: list[_PendingChunk] = []
        self._runs: list[RunInsert] = []
        self._steps: list[_PendingStep] = []
        self._actions: list[_PendingAction] = []
        self._optional_backfills: list[OptionalProjectionTask] = []
        self._vec_inserts: list[_PendingVec] = []
        self._operational_writes: list[OperationalAppend | OperationalPut | OperationalDelete] = []

    def add_node(
        self,
        *,
        row_id: str,
        logical_id: str,
        kind: str,
        properties: Any,
        source_ref: str | None = None,
        upsert: bool = False,
        chunk_policy: ChunkPolicy = ChunkPolicy.PRESERVE,
        content_ref: str | None = None,
    ) -> NodeHandle:
        """Add a node insert to the write request.

        Args:
            row_id: Unique row identifier for this version of the node.
            logical_id: Stable logical identifier for the node.
            kind: Node kind (type label).
            properties: JSON-serializable properties payload.
            source_ref: Optional provenance source reference.
            upsert: If True, replace an existing node with the same logical ID.
            chunk_policy: How to handle existing chunks on upsert.
            content_ref: Optional URI referencing external content.

        Returns
        -------
            A NodeHandle that can be used to reference this node elsewhere
            in the same builder.
        """
        handle = NodeHandle(row_id=row_id, logical_id=logical_id, _builder_id=self._builder_id)
        self._nodes.append(
            NodeInsert(
                row_id=row_id,
                logical_id=logical_id,
                kind=kind,
                properties=properties,
                source_ref=source_ref,
                upsert=upsert,
                chunk_policy=chunk_policy,
                content_ref=content_ref,
            )
        )
        return handle

    def retire_node(self, *, logical_id: NodeHandle | str, source_ref: str | None = None) -> None:
        """Mark a node as retired (soft-delete) by logical ID or handle."""
        self._node_retires.append(_PendingNodeRetire(logical_id=logical_id, source_ref=source_ref))

    def add_edge(
        self,
        *,
        row_id: str,
        logical_id: str,
        source: NodeHandle | str,
        target: NodeHandle | str,
        kind: str,
        properties: Any,
        source_ref: str | None = None,
        upsert: bool = False,
    ) -> EdgeHandle:
        """Add an edge insert connecting two nodes.

        Args:
            row_id: Unique row identifier for this version of the edge.
            logical_id: Stable logical identifier for the edge.
            source: Source node (handle or logical ID string).
            target: Target node (handle or logical ID string).
            kind: Edge kind (type label).
            properties: JSON-serializable properties payload.
            source_ref: Optional provenance source reference.
            upsert: If True, replace an existing edge with the same logical ID.

        Returns
        -------
            An EdgeHandle that can be used to reference this edge elsewhere
            in the same builder.
        """
        handle = EdgeHandle(logical_id=logical_id, _builder_id=self._builder_id)
        self._edges.append(
            _PendingEdge(
                row_id=row_id,
                logical_id=logical_id,
                source=source,
                target=target,
                kind=kind,
                properties=properties,
                source_ref=source_ref,
                upsert=upsert,
            )
        )
        return handle

    def retire_edge(self, *, logical_id: EdgeHandle | str, source_ref: str | None = None) -> None:
        """Mark an edge as retired (soft-delete) by logical ID or handle."""
        self._edge_retires.append(_PendingEdgeRetire(logical_id=logical_id, source_ref=source_ref))

    def add_chunk(
        self,
        *,
        id: str,
        node: NodeHandle | str,
        text_content: str,
        byte_start: int | None = None,
        byte_end: int | None = None,
        content_hash: str | None = None,
    ) -> ChunkHandle:
        """Add a text chunk associated with a node.

        Args:
            id: Unique chunk identifier.
            node: Owning node (handle or logical ID string).
            text_content: The text content of the chunk.
            byte_start: Optional byte offset where the chunk starts in the source.
            byte_end: Optional byte offset where the chunk ends in the source.
            content_hash: Optional hash of the external content this chunk was derived from.

        Returns
        -------
            A ChunkHandle for referencing this chunk in vector inserts.
        """
        handle = ChunkHandle(id=id, _builder_id=self._builder_id)
        self._chunks.append(
            _PendingChunk(
                id=id,
                node=node,
                text_content=text_content,
                byte_start=byte_start,
                byte_end=byte_end,
                content_hash=content_hash,
            )
        )
        return handle

    def add_run(
        self,
        *,
        id: str,
        kind: str,
        status: str,
        properties: Any,
        source_ref: str | None = None,
        upsert: bool = False,
        supersedes_id: str | None = None,
    ) -> RunHandle:
        """Add a run insert to the write request.

        Returns
        -------
            A RunHandle for referencing this run when adding steps.
        """
        handle = RunHandle(id=id, _builder_id=self._builder_id)
        self._runs.append(
            RunInsert(
                id=id,
                kind=kind,
                status=status,
                properties=properties,
                source_ref=source_ref,
                upsert=upsert,
                supersedes_id=supersedes_id,
            )
        )
        return handle

    def add_step(
        self,
        *,
        id: str,
        run: RunHandle | str,
        kind: str,
        status: str,
        properties: Any,
        source_ref: str | None = None,
        upsert: bool = False,
        supersedes_id: str | None = None,
    ) -> StepHandle:
        """Add a step insert belonging to a run.

        Returns
        -------
            A StepHandle for referencing this step when adding actions.
        """
        handle = StepHandle(id=id, _builder_id=self._builder_id)
        self._steps.append(
            _PendingStep(
                id=id,
                run=run,
                kind=kind,
                status=status,
                properties=properties,
                source_ref=source_ref,
                upsert=upsert,
                supersedes_id=supersedes_id,
            )
        )
        return handle

    def add_action(
        self,
        *,
        id: str,
        step: StepHandle | str,
        kind: str,
        status: str,
        properties: Any,
        source_ref: str | None = None,
        upsert: bool = False,
        supersedes_id: str | None = None,
    ) -> ActionHandle:
        """Add an action insert belonging to a step.

        Returns
        -------
            An ActionHandle for referencing this action.
        """
        handle = ActionHandle(id=id, _builder_id=self._builder_id)
        self._actions.append(
            _PendingAction(
                id=id,
                step=step,
                kind=kind,
                status=status,
                properties=properties,
                source_ref=source_ref,
                upsert=upsert,
                supersedes_id=supersedes_id,
            )
        )
        return handle

    def add_optional_backfill(self, target: ProjectionTarget | str, payload: Any) -> None:
        """Queue an optional projection backfill task (e.g. FTS or vector)."""
        value = target if isinstance(target, ProjectionTarget) else ProjectionTarget(target)
        self._optional_backfills.append(OptionalProjectionTask(target=value, payload=payload))

    def add_vec_insert(self, *, chunk: ChunkHandle | str, embedding: list[float]) -> None:
        """Add a vector embedding associated with a chunk."""
        self._vec_inserts.append(_PendingVec(chunk=chunk, embedding=embedding))

    def add_operational_append(
        self,
        *,
        collection: str,
        record_key: str,
        payload_json: Any,
        source_ref: str | None = None,
    ) -> None:
        """Append a mutation to an operational collection."""
        self._operational_writes.append(
            OperationalAppend(
                collection=collection,
                record_key=record_key,
                payload_json=payload_json,
                source_ref=source_ref,
            )
        )

    def add_operational_put(
        self,
        *,
        collection: str,
        record_key: str,
        payload_json: Any,
        source_ref: str | None = None,
    ) -> None:
        """Put (upsert) a record into an operational collection."""
        self._operational_writes.append(
            OperationalPut(
                collection=collection,
                record_key=record_key,
                payload_json=payload_json,
                source_ref=source_ref,
            )
        )

    def add_operational_delete(
        self,
        *,
        collection: str,
        record_key: str,
        source_ref: str | None = None,
    ) -> None:
        """Delete a record from an operational collection."""
        self._operational_writes.append(
            OperationalDelete(
                collection=collection,
                record_key=record_key,
                source_ref=source_ref,
            )
        )

    def build(self) -> WriteRequest:
        """Resolve all handles and produce a finalized WriteRequest.

        Raises
        ------
            BuilderValidationError: If any handle belongs to a different builder.
        """
        return WriteRequest(
            label=self._label,
            nodes=list(self._nodes),
            node_retires=[
                NodeRetire(
                    logical_id=self._resolve_node_ref(item.logical_id),
                    source_ref=item.source_ref,
                )
                for item in self._node_retires
            ],
            edges=[
                EdgeInsert(
                    row_id=item.row_id,
                    logical_id=item.logical_id,
                    source_logical_id=self._resolve_node_ref(item.source),
                    target_logical_id=self._resolve_node_ref(item.target),
                    kind=item.kind,
                    properties=item.properties,
                    source_ref=item.source_ref,
                    upsert=item.upsert,
                )
                for item in self._edges
            ],
            edge_retires=[
                EdgeRetire(
                    logical_id=self._resolve_edge_ref(item.logical_id),
                    source_ref=item.source_ref,
                )
                for item in self._edge_retires
            ],
            chunks=[
                ChunkInsert(
                    id=item.id,
                    node_logical_id=self._resolve_node_ref(item.node),
                    text_content=item.text_content,
                    byte_start=item.byte_start,
                    byte_end=item.byte_end,
                    content_hash=item.content_hash,
                )
                for item in self._chunks
            ],
            runs=list(self._runs),
            steps=[
                StepInsert(
                    id=item.id,
                    run_id=self._resolve_run_ref(item.run),
                    kind=item.kind,
                    status=item.status,
                    properties=item.properties,
                    source_ref=item.source_ref,
                    upsert=item.upsert,
                    supersedes_id=item.supersedes_id,
                )
                for item in self._steps
            ],
            actions=[
                ActionInsert(
                    id=item.id,
                    step_id=self._resolve_step_ref(item.step),
                    kind=item.kind,
                    status=item.status,
                    properties=item.properties,
                    source_ref=item.source_ref,
                    upsert=item.upsert,
                    supersedes_id=item.supersedes_id,
                )
                for item in self._actions
            ],
            optional_backfills=list(self._optional_backfills),
            vec_inserts=[
                VecInsert(
                    chunk_id=self._resolve_chunk_ref(item.chunk),
                    embedding=item.embedding,
                )
                for item in self._vec_inserts
            ],
            operational_writes=list(self._operational_writes),
        )

    def _resolve_node_ref(self, value: NodeHandle | str) -> str:
        if isinstance(value, NodeHandle):
            if value._builder_id != self._builder_id:
                raise BuilderValidationError(
                    "node handle belongs to a different WriteRequestBuilder"
                )
            return value.logical_id
        return value

    def _resolve_edge_ref(self, value: EdgeHandle | str) -> str:
        if isinstance(value, EdgeHandle):
            if value._builder_id != self._builder_id:
                raise BuilderValidationError(
                    "edge handle belongs to a different WriteRequestBuilder"
                )
            return value.logical_id
        return value

    def _resolve_run_ref(self, value: RunHandle | str) -> str:
        if isinstance(value, RunHandle):
            if value._builder_id != self._builder_id:
                raise BuilderValidationError(
                    "run handle belongs to a different WriteRequestBuilder"
                )
            return value.id
        return value

    def _resolve_step_ref(self, value: StepHandle | str) -> str:
        if isinstance(value, StepHandle):
            if value._builder_id != self._builder_id:
                raise BuilderValidationError(
                    "step handle belongs to a different WriteRequestBuilder"
                )
            return value.id
        return value

    def _resolve_chunk_ref(self, value: ChunkHandle | str) -> str:
        if isinstance(value, ChunkHandle):
            if value._builder_id != self._builder_id:
                raise BuilderValidationError(
                    "chunk handle belongs to a different WriteRequestBuilder"
                )
            return value.id
        return value
